### PR TITLE
[Keyboard] Add UM-A Keyboard

### PR DESCRIPTION
**Pull request by duoshock**
## Description

Please add UM-A keyboard - Alice style split keyboard with encoder and OLED screen.
https://mergedesign.store/products/um-a

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/merge/uma/info.json [hunk 1]
```diff
@@ -0,0 +1,21 @@
+{
+    "keyboard_name": "UM-A",
+    "url": "https://mergedesign.store/products/um-a",
+    "maintainer": "duoshock",
+    "width": 18.75,
+    "height": 5.25,
```
Comment (drashna) [inline keyboards/merge/uma/info.json:6]:
```suggestion
```

File: keyboards/merge/uma/keymaps/default/keymap.c [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include QMK_KEYBOARD_H
+#include <stdio.h>
```
Comment (drashna) [inline keyboards/merge/uma/keymaps/default/keymap.c:18]:
```suggestion
```

File: keyboards/merge/uma/keymaps/default/keymap.c [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+    char wpm_str[8];
+    oled_set_cursor(0,13);
+    sprintf(wpm_str, " %03d", current_wpm);
+    oled_write(wpm_str, false);
+    oled_set_cursor(0,14);
+    oled_write(" WPM ", false);
+    // WPM counter End
+}
+
+bool oled_task_user(void) {
```
Comment (drashna) [inline keyboards/merge/uma/keymaps/default/keymap.c:172]: could y move the oled code into the keyboards c file?

File: keyboards/merge/uma/keymaps/default/keymap.c [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+    //oled_write_ln_P(PSTR(" "), false);
+    oled_write_P(PSTR("-----"), false);
+
+    // WPM counter Start (Need #include <stdio.h> to work)
+    char wpm_str[8];
+    oled_set_cursor(0,13);
+    sprintf(wpm_str, " %03d", current_wpm);
+    oled_write(wpm_str, false);
+    oled_set_cursor(0,14);
+    oled_write(" WPM ", false);
```
Comment (drashna) [inline keyboards/merge/uma/keymaps/default/keymap.c:168]: https://docs.qmk.fm/#/squeezing_avr?id=oled-tweaks

File: keyboards/merge/uma/rules.mk [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+CONSOLE_ENABLE = no         # Console for debug
+COMMAND_ENABLE = no         # Commands for debug and configuration
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+NKRO_ENABLE = yes           # USB Nkey Rollover
+BACKLIGHT_ENABLE = yes       # Enable keyboard backlight functionality
+RGBLIGHT_ENABLE = no       # Enable keyboard RGB underglow
+BLUETOOTH_ENABLE = no       # Enable Bluetooth
+AUDIO_ENABLE = no           # Audio output
```
Comment (drashna) [inline keyboards/merge/uma/rules.mk:22]:
```suggestion
NKRO_ENABLE = yes           # USB Nkey Rollover
BACKLIGHT_ENABLE = yes       # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = no       # Enable keyboard RGB underglow
AUDIO_ENABLE = no           # Audio output
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/merge/uma/keymaps/default/keymap.c (hunk 3 above)
Comment (duoshock) [reply keyboards/merge/uma/keymaps/default/keymap.c]: Sure. I've moved it to the keyboards c file.

File: keyboards/merge/uma/keymaps/default/keymap.c (hunk 4 above)
Comment (duoshock) [reply keyboards/merge/uma/keymaps/default/keymap.c]: Thanks for this. It's definitely saved a good amount of memory!

File: keyboards/merge/uma/readme.md [hunk 6]
```diff
@@ -0,0 +1,14 @@
+# UM-70
+
+![UM-A](https://i.imgur.com/Bm1mRDbl.jpg)
+
+A Alice like layout keyboard with arrow keys, encoder and OLED screen sold by Merge. [Product page](https://mergedesign.store/products/um-70)
+
+* Keyboard Maintainer: [duoshock](https://github.com/duoshock)
+* Hardware Availability: [Merge Store](https://mergedesign.store/products/um-a)
+
+Make example for this keyboard (after setting up your build environment):
```
Comment (drashna) [inline keyboards/merge/uma/readme.md:10]: Could you add instructions on how to enter the bootloader, to flash new firmware? 

Eg: 
```markup
## Bootloader

Enter the bootloader in 3 ways:

* **Bootmagic reset**: Hold down the key at (0,0) in the matrix (usually the top left key or Escape) and plug in the keyboard
* **Physical reset button**: Briefly press the button on the back of the PCB - some may have pads you must short instead
* **Keycode in layout**: Press the key mapped to `RESET` if it is available
```

File: keyboards/merge/uma/keymaps/via/keymap.c [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        }
+    } else if (index == 1) { /* Master Right */
+        if (clockwise) {
+            tap_code(KC_VOLU);
+        } else {
+            tap_code(KC_VOLD);
+        }
+    }
+    return true;
+}
```
Comment (drashna) [inline keyboards/merge/uma/keymaps/via/keymap.c:81]:
```suggestion
```

File: keyboards/merge/uma/keymaps/default/keymap.c [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+        }
+    } else if (index == 1) { /* Master Right */
+        if (clockwise) {
+            tap_code(KC_VOLU);
+        } else {
+            tap_code(KC_VOLD);
+        }
+    }
+    return true;
+}
```
Comment (drashna) [inline keyboards/merge/uma/keymaps/default/keymap.c:81]:
```suggestion
```

File: keyboards/merge/uma/uma.c [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+ */
+
+#include "uma.h"
+
+#ifdef OLED_ENABLE
+void suspend_power_down_user(void) {
+    oled_off();
+}
+
+oled_rotation_t oled_init_user(oled_rotation_t rotation) {
```
Comment (drashna) [inline keyboards/merge/uma/uma.c:24]:
```suggestion
oled_rotation_t oled_init_kb(oled_rotation_t rotation) {
```

File: keyboards/merge/uma/uma.c [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+    // WPM counter Start
+    oled_set_cursor(0,13);
+    oled_write(" ", false);
+    oled_write(get_u8_str(get_current_wpm(), '0'), false);
+    oled_set_cursor(0,14);
+    oled_write(" WPM ", false);
+    // WPM counter End
+}
+
+bool oled_task_user(void) {
```
Comment (drashna) [inline keyboards/merge/uma/uma.c:98]:
```suggestion
bool oled_task_kb(void) {
    if (!oled_task_user()) { return false; }
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/merge/uma/readme.md (hunk 6 above)
Comment (duoshock) [reply keyboards/merge/uma/readme.md:10]: Sure thing. I've just updated it. Thanks!

File: keyboards/merge/uma/config.h [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+#define UNUSED_PINS
+
+#define BACKLIGHT_PIN C6
+
+/* COL2ROW or ROW2COL */
+#define DIODE_DIRECTION COL2ROW
+
+#define SOFT_SERIAL_PIN D2
+#define SPLIT_USB_DETECT
+#define MASTER_LEFT
```
Comment (drashna) [inline keyboards/merge/uma/config.h:39]: This is the default behavior and doesnt need to be defined
```suggestion
```

File: keyboards/merge/uma/uma.c [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+    oled_write(" ", false);
+    oled_write(get_u8_str(get_current_wpm(), '0'), false);
+    oled_set_cursor(0,14);
+    oled_write(" WPM ", false);
+    // WPM counter End
+}
+
+bool oled_task_kb(void) {
+    if (!oled_task_user()) { return false; }
+    current_wpm = get_current_wpm();
```
Comment (drashna) [inline keyboards/merge/uma/uma.c:100]:
```suggestion
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/merge/uma/rules.mk [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+
+# Build Options
+#   change yes to no to disable
+#
+BOOTMAGIC_ENABLE = yes     # Enable Bootmagic Lite
+MOUSEKEY_ENABLE = yes       # Mouse keys
+EXTRAKEY_ENABLE = yes       # Audio control and System control
+CONSOLE_ENABLE = no         # Console for debug
+COMMAND_ENABLE = no         # Commands for debug and configuration
+NKRO_ENABLE = yes           # USB Nkey Rollover
```
Comment (fauxpark) [inline keyboards/merge/uma/rules.mk:15]:
```suggestion
NKRO_ENABLE = yes           # Enable N-Key Rollover
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Review by fauxpark — APPROVED** [review]

File: keyboards/merge/uma/keymaps/default/rules.mk [hunk 14]
```diff
@@ -0,0 +1,2 @@
+WPM_ENABLE = yes
+LTO_ENABLE = yes
```
Comment (drashna) [inline keyboards/merge/uma/keymaps/default/rules.mk:2]:
```suggestion
```

File: keyboards/merge/uma/keymaps/via/rules.mk [hunk 15]
```diff
@@ -0,0 +1,3 @@
+VIA_ENABLE = yes
+WPM_ENABLE = yes
+LTO_ENABLE = yes
```
Comment (drashna) [inline keyboards/merge/uma/keymaps/via/rules.mk:3]:
```suggestion
```

File: keyboards/merge/uma/rules.mk [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+EXTRAKEY_ENABLE = yes       # Audio control and System control
+CONSOLE_ENABLE = no         # Console for debug
+COMMAND_ENABLE = no         # Commands for debug and configuration
+NKRO_ENABLE = yes           # Enable N-Key Rollover
+BACKLIGHT_ENABLE = yes       # Enable keyboard backlight functionality
+RGBLIGHT_ENABLE = no       # Enable keyboard RGB underglow
+AUDIO_ENABLE = no           # Audio output
+ENCODER_ENABLE = yes
+SPLIT_KEYBOARD = yes
+OLED_ENABLE = yes
```
Comment (drashna) [inline keyboards/merge/uma/rules.mk:21]:
```suggestion
LTO_ENABLE = yes
OLED_ENABLE = yes
```

**Review by drashna — CHANGES_REQUESTED** [review]
If LTO needs to be enabled on each of the keymaps, it would be better to enable it at the keyboard level, in general.

**Review by drashna — APPROVED** [review]

**Comment by drashna** [comment]
Thanks!

**Pull request merged**